CDCgov/prime-simplereport
# Patient bulk upload email notification

# BACKEND PULL REQUEST

## Related Issue #4440 

Adds an email notification once patient bulk upload is complete. The email includes a link back to the application, where if logged in, the user can view the newly uploaded patients.

⚠️ **Note that this is built on top of branch `emma/async-3`. and cannot be merged until that branch is merged. I'll leave this PR as a draft until the async behavior for patient bulk upload is merged into main.**

💡 **[This compare link](https://github.com/CDCgov/prime-simplereport/compare/emma/async-3...george/4440-patient-upload-email) will probably be more useful to view changes**

## Changes Proposed

-  adds an email template for patient bulk upload
- when save to the database save in the async method for patient bulk upload is complete, an email is sent to notify the user
- if processing fails, an email is sent to notify the user (screenshot [here](https://github.com/CDCgov/prime-simplereport/issues/4440#issuecomment-1324262939))

## Additional Information

- I added one unit test for the new email template
- There should probably be another test to verify that the success email gets sent in the async method, however this will be difficult to write until the changes for tests are completed on `emma/async-3`. I'll update this when we have a clearer testing strategy for the async method. Backend tests are expected to fail right now.
- I'll also need to to add a unit test for the error email behavior, if processing within the while loop fails, or if the saveAll to the db fails.

Here is the sequence of events: before patient upload, after patient upload, and then the email that gets sent:
- before
![dev4 b4](https://user-images.githubusercontent.com/89797785/203332989-95ca9ecf-5622-4393-a452-1315b4571431.png)
- after
![dev4 after](https://user-images.githubusercontent.com/89797785/203333015-6dc14a18-7291-4295-838d-ef0702d06cc3.png)
- email
![dev4 email](https://user-images.githubusercontent.com/89797785/203333030-5d2f0a74-e159-40ec-ac12-88457bf8f336.png)

When an error occurs in processing, sending the error email is triggered. This isn't manually testable in an environment, so here are the logs when I triggered errors in the code:
- processing error in while loop (turned off validation in `PatientBulkUploadService` and uploaded an invalid csv
![Screen Shot 2022-11-22 at 4 01 21 PM](https://user-images.githubusercontent.com/89797785/203425065-280e22fe-cd76-43c7-8f65-a815f4df5672.png)
- processing error in saveAll (forced an error in the `addPatientsAndPhoneNumbers` method)
![Screen Shot 2022-11-22 at 4 15 31 PM](https://user-images.githubusercontent.com/89797785/203425096-29174e35-830c-471c-846c-680be6f6480a.png)

## Testing

How should reviewers verify this PR?
- Go to dev4, upload patients via csv, and confirm that you receive an email notifying you when the upload is complete. Click the `See patients` link, which should take you directly to the patients page, if you are logged in to dev4.

## Checklist for Primary Reviewer
